### DISCUSSION TITLE
Example of input constraints

### DISCUSSION THREAD

**Discussion by BSNunes** [meta-pytorch/botorch#1589] [Q&A]
I am working on optimizing a function with multiple input parameters, f(x1, x2, ..., xn) and, in order to verify if the optimal parameters found are not a local maximum, I would like to impose some input constraints, in order to force the algorithm to look for another set of optimal parameters.

To illustrate the situation, I wrote a simple code (copied below), aiming to optimize the function f(x,y) = cos(x) * sin(y), where -6 < x, y < 6. This function has ten local maxima within this range, and the algorithm converges to one of them very quickly. Hence, I would like to add a restriction on x and y near this maximum, in order to make botorch search for another maximum. Could you please point me some example (I find a outcome constraints example https://botorch.org/tutorials/closed_loop_botorch_only, but I did not find any simple parameter constraints example on botorch documentation page), or help me with this matter? Any help will be greatly appreciated. Thank you for your attention.

```
import torch
import random as rn
import numpy as np
from botorch.models import SingleTaskGP
from gpytorch.mlls import ExactMarginalLogLikelihood
from botorch.fit import fit_gpytorch_model
from botorch.acquisition import UpperConfidenceBound
from botorch.optim import optimize_acqf

#function to be optimize
def F(x,y):
    f = np.cos(x)*np.cos(y)
    return(f)

#find next points to evaluate the function
def get_next_points(beta,init_x, init_y, bounds, N_point=1):
    single_model = SingleTaskGP(init_x, init_y)
    mll = ExactMarginalLogLikelihood(single_model.likelihood, single_model)
    fit_gpytorch_model(mll)
    UCB = UpperConfidenceBound(single_model, beta=beta)
    candidates, acq_value = optimize_acqf(
                acq_function=UCB,
                bounds=bounds,
                q=N_point,
                num_restarts=200,
                raw_samples=1024,
                options={"batch_limit": 5, "maxiter": 200})
    return candidates

#range parameters and initial points
xi = -6.
xf = 6.
yi = -6.
yf = 6.
n_points=10
N_point=1
beta=0.2
bounds = torch.DoubleTensor([[xi,yi],[xf,yf]]) 
for i in range(n_points):
    x=rn.uniform(xi,xf)
    y=rn.uniform(yi,yf)
    if i == 0:
        X_train = torch.DoubleTensor([[x,y]])
        Y = torch.DoubleTensor([[F(x,y)]])
    else:
        X_train1 = torch.DoubleTensor([[x,y]])
        Y1 = torch.DoubleTensor([[F(x,y)]])
        Y=torch.cat([Y,Y1])
        X_train=torch.cat([X_train,X_train1])

#optimization loop
for i in range(30):
    new_candidates =  get_next_points(beta,X_train, Y, bounds,N_point)
    x= new_candidates[0,0].item()
    y= new_candidates[0,1].item()

    X_train1 = torch.DoubleTensor([[x,y]])
    Y1 = torch.DoubleTensor([[F(x,y)]])
    Y=torch.cat([Y,Y1])
    X_train=torch.cat([X_train,X_train1])
    
    #print results
    print(i,x,y,F(x,y))
```

**Comment by Balandat** [comment]
`optimize_acqf` takes linear inequality (and also equality) constraints, the docstring explains how to pass these in: https://github.com/pytorch/botorch/blob/main/botorch/optim/optimize.py#L83-L88

> I would like to add a restriction on x and y near this maximum

Are you planning to "cut out" some region around one maximum? Note that this would be a non-convex constraint and so couldn't be expressed by a set of linear inequality constraints. In that case you'd have to pass in a nonlinear inequality constraint: https://github.com/pytorch/botorch/blob/main/botorch/optim/optimize.py#L89-L96. You'll also have to pass in your own initial conditions though (related discussion in #1572).

**Reply by BSNunes** [reply]
Thank you.
Yes, exactly. I would like to manually cut out a small region close to this maximum based on the results after a certain number of iterations.
I will try to code the nonlinear inequality constraint.

**Comment by eytan** [comment]
It is surprising that the algorithm wouldn’t explore the other minima after
discovering one of the others. Have you tried running for more iterations
or using a better acquisition function like EI?  It would be surprising if
you had to code in additional constraints here.

**Reply by BSNunes** [reply]
I do not know if there are any other maximum in my problem, but I do suppose they exist. It is a truly black-box function. The algorithm run over 300 iterations and after converging to one point it stayed there, without looking for others. 

I haven't tried any other acquisition function. After starting with one would be strange to change it.

With this constraint, I would like to confirm this is the only maximum (or the higher one) in my parameters domain.

**Reply by Balandat** [reply]
As part of the optimization the algorithm should already have explored other parts of the search space. As a first step I would recommend verifying that the cross-validated model fits (https://botorch.org/tutorials/batch_mode_cross_validation) look good across the domain; if that's the case then this is a reasonable indicator that the algorithm has found the global optimum.

**Comment by eytan** [comment]
Are you just using botorch for black box optimization or are you
specifically looking to develop your own algorithms for BO? If it’s the
former you may want to check out ax http://ax.dev . It uses botorch
underneath the hood and does a lot of transformations underneath the hood
to make sure everything is setup well.

**Reply by BSNunes** [reply]
I'm using it just for black box optimization. It has been as major tool for my research.
